src/protocol/ContactRequestChannel.cpp

```diff
@@ -0,0 +1,305 @@
+    // If the channel is already open, the response is sent as a separate packet
+    if (isOpened()) {
+        Response response;
+        response.set_status(m_responseStatus);
+        if (!m_responseErrorMessage.isEmpty())
+            response.set_error_message(m_responseErrorMessage.toStdString());
+        sendMessage(response);
+
+        if (m_responseStatus > Response::Pending)
+            closeChannel();
+    }
+}
+
+bool ContactRequestChannel::allowOutboundChannelRequest(Data::Control::OpenChannel *request)
+{
+    if (connection()->direction() != Connection::ClientSide ||
+        connection()->purpose() != Connection::Purpose::OutboundRequest)
+    {
+        BUG() << "ContactRequestChannel can only be used on OutboundRequest connections. Has purpose"
+              << int(connection()->purpose());
+        return false;
+    }
+
+    if (connection()->findChannel<ContactRequestChannel>()) {
+        BUG() << "ContactRequestChannel can only be used once per connection";
+        return false;
+    }
+
+    QScopedPointer<Data::ContactRequest::ContactRequest> contactData(new Data::ContactRequest::ContactRequest);
+    if (!m_nickname.isEmpty())
+        contactData->set_nickname(m_nickname.toStdString());
+    if (!m_message.isEmpty())
+        contactData->set_message_text(m_message.toStdString());
+
+    request->SetAllocatedExtension(Data::ContactRequest::contact_request, contactData.take());
+    return true;
+}
+
+bool ContactRequestChannel::processChannelOpenResult(const Data::Control::ChannelResult *result)
+{
+    if (!result->HasExtension(Data::ContactRequest::response)) {
+        qDebug() << "Expected a response for the contact request";
+        return false;
+    }
+
+    Data::ContactRequest::Response response = result->GetExtension(Data::ContactRequest::response);
+    return handleResponse(&response);
+}
+
+void ContactRequestChannel::receivePacket(const QByteArray &packet)
+{
+    Data::ContactRequest::Response response;
+    if (!response.ParseFromArray(packet.constData(), packet.size())) {
+        closeChannel();
+        return;
+    }
+
+    if (!handleResponse(&response))
+        closeChannel();
+}
+
+bool ContactRequestChannel::handleResponse(const Data::ContactRequest::Response *response)
+{
+    using namespace Data::ContactRequest;
+    if (response->status() == Response::Undefined) {
+        qDebug() << "Got an invalid response (undefined status) to a contact request";
+        return false;
+    }
+
+    if (m_responseStatus > Response::Pending) {
+        qDebug() << "Received a response" << response->status() << "to a contact request which already had a final response" << m_responseStatus;
+        return false;
+    }
+
+    QString message(QString::fromStdString(response->error_message()));
+    if (message.size() > MessageMaxCharacters) {
+        qDebug() << "Received a contact request response with an overly long message of " << message.size() << "characters";
+        message.clear();
+    }
```

Comment:
It's a response error message; we can't refuse the response without making everyone confused.

I think this error message is removed in a later set of patches, in favor of an enum.